crates/qpy/src/circuit_writer.rs

```diff
@@ -0,0 +1,1144 @@
+                    match getattr_or_none(gate, "definition") {
+                        None => (),
+                        Some(definition) => {
+                            has_definition = true;
+                            data = serialize(&pack_circuit(
+                                &mut definition.extract()?,
+                                Some(py.None().bind(py)),
+                                false,
+                                qpy_data.version,
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                OperationRef::Instruction(pyinst) => {
+                    let inst = pyinst.instruction.bind(py);
+                    match getattr_or_none(inst, "definition") {
+                        None => (),
+                        Some(definition) => {
+                            has_definition = true;
+                            data = serialize(&pack_circuit(
+                                &mut definition.extract()?,
+                                Some(py.None().bind(py)),
+                                false,
+                                qpy_data.version,
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                OperationRef::Operation(pyoperation) => {
+                    let operation = pyoperation.operation.bind(py);
+                    match getattr_or_none(operation, "definition") {
+                        None => (),
+                        Some(definition) => {
+                            has_definition = true;
+                            data = serialize(&pack_circuit(
+                                &mut definition.extract()?,
+                                Some(py.None().bind(py)),
+                                false,
+                                qpy_data.version,
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+        let num_qubits = operation.num_qubits();
+        let num_clbits = operation.num_clbits();
+        if !base_gate.is_none() {
+            let op_parts = base_gate.extract::<OperationFromPython>()?;
+            let instruction = CircuitInstruction {
+                operation: op_parts.operation,
+                qubits: PyTuple::empty(py).into(),
+                clbits: PyTuple::empty(py).into(),
+                params: op_parts.params,
+                label: op_parts.label,
+                #[cfg(feature = "cache_pygates")]
+                py_op: std::sync::OnceLock::new(),
+            };
+            let packed_instruction = qpy_data.circuit_data.pack(py, &instruction)?;
```

Comment:
Can we avoid this call and just create a `PackedInstruction` instance directly? It's a bit weird that that writer has a side effect on `CircuitData`...